common/src/patch.rs

```diff
@@ -0,0 +1,219 @@
+    match git::add_tag(repo, message, &patch_tag_name) {
+        Ok(_) => {
+            if verbose {
+                spinner.finish();
+            }
+        }
+        Err(err) => {
+            spinner.failed();
+            return Err(err);
+        }
+    };
+
+    match git::push_tag(&patch_tag_name) {
+        Ok(output) => {
+            if verbose {
+                term::blob(output);
+            }
+        }
+        Err(err) => {
+            spinner.failed();
+            return Err(err);
+        }
+    };
+
+    spinner.message("Setting up branch...".to_owned());
+    match git::push_branch(current_branch) {
+        Ok(output) => {
+            if verbose {
+                spinner.finish();
+                term::blob(output);
+            }
+        }
+        Err(err) => {
+            spinner.failed();
+            return Err(err);
+        }
+    };
+
+    spinner.finish();
+
+    Ok(())
+}
+
+/// Adds patch details as a new row to `table` and render later.
+pub fn print<S>(storage: &S, patch: &Metadata, table: &mut term::Table<2>) -> anyhow::Result<()>
+where
+    S: AsRef<ReadOnly>,
+{
+    let storage = storage.as_ref();
+
+    if let Some(message) = patch.message.clone() {
+        let you = patch.peer.id == *storage.peer_id();
+        let title = message.lines().next().unwrap_or("");
+        let name = term::format::tertiary(&patch.name());
+
+        let mut author_info = vec![term::format::italic(format!(
+            "└── Opened by {}",
+            &patch.peer.name.as_ref().unwrap_or(&"".to_owned())
+        ))];
+
+        if you {
+            author_info.push(term::format::badge_secondary("you"));
+        }
+
+        table.push([term::format::bold(title), "".to_owned()]);
+        table.push([author_info.join(" "), name]);
+    }
+    Ok(())
+}
+
+/// Tries to construct a patch from ['git2::Tag'] and ['project::PeerInfo'].
+/// If the tag name matches the radicle patch prefix, a new patch metadata is
+/// created.
+pub fn from_tag(tag: git2::Tag, info: project::PeerInfo) -> Result<Option<Metadata>> {
+    let patch = match tag.name() {
+        Some(id) => {
+            let prefix = Regex::new(&format!("{}*", TAG_PREFIX)).unwrap();
+            if prefix.is_match(id) {
+                Some(Metadata {
+                    id: id.to_string(),
+                    peer: info,
+                    message: tag.message().map(|m| m.to_string()),
+                    commit: tag.target_id(),
+                })
+            } else {
+                None
+            }
+        }
+        None => None,
+    };
+
+    Ok(patch)
+}
+
+pub fn self_info<S>(storage: &S, project: &project::Metadata) -> Result<project::PeerInfo>
+where
+    S: AsRef<ReadOnly>,
+{
+    let storage = storage.as_ref();
+    let urn = storage.config()?.user()?.unwrap();
+    let peer_id = storage.peer_id();
+    let name = person::get(storage, &urn)?.map(|p| p.subject().name.to_string());
```

Comment:
Calling `person::local` needs a non-`ReadOnly` storage. And since passing a `ReadOnly` storage into `self_info` makes sense semantically, I'd lean towards leaving the code as is. Otherwise we'd need to get the writeable storage using a signer etc, which would result in even more code, I guess.

Reply:
That's a bug in `person::local` then I guess, but let's solve it another day.